lib/logstash/inputs/redis.rb

```diff
@@ -184,15 +279,64 @@ def list_listener(redis, output_queue)
     end
   end
 
-  private
-  def channel_listener(redis, output_queue)
-    redis.subscribe @key do |on|
+  # private
+  def subscribe_teardown
+    return if @redis.nil? || !@redis.connected?
+
+    # the underlying client in redis is blocked on subscribe
+    # need new redis instance, send poison message, causes
+    # an unsubscribe
+    publish_poison_message
+
+    # if its a SubscribedClient then:
+    # a) the poison_message did not work
+    # b) it does not have a disconnect method (yet)
+    if @redis.client.is_a?(::Redis::Client)
+      @redis.client.disconnect
+    end
+    @redis = nil
+  end
+
+  # private
+  def publish_poison_message
+    new_redis_instance.publish(@key, REDIS_INPUT_POISON_MSG)
```

Comment:
Is writing through redis the only way to do this? Is there no way to interrupt the redis subscriber?